torch/nn/functional.py

```diff
@@ -571,6 +571,37 @@ def batch_norm(input, running_mean, running_var, weight=None, bias=None,
     return f(input, weight, bias)
 
 
+def relation(input, g, f=None, embedding=None, max_pairwise=None):
+    # Batch size, number of objects, feature size
+    b, o, c = input.size()
+    # Construct pairwise indices
+    i = Variable(torch.arange(0, o).type_as(input.data).long().repeat(o))
+    j = Variable(torch.arange(0, o).type_as(input.data).long().repeat(o, 1).t().contiguous().view(-1))
+    # Create pairwise matrix
+    pairs = torch.cat((torch.index_select(input, 1, i), torch.index_select(input, 1, j)), 2)
```

Comment:
To avoid allocating new tensors `i` and `j`, I think you can directly write:
```
pairs = torch.cat((input.unsqueeze(1).expand(b, o, o, c).view(b, o*o, c),
                   input.unsqueeze(2).expand(b, o, o, c).view(b, o*o, c), 2)
```